Optimize AAE replication.

cmeiklejohn opened the pull request:
Implement a prototype optimization for active anti-entropy replication which front-loads the first two levels of bucket retrieval through one TCP request, instead of 1025 roundtrips.  Futher optimizations include increasing the number of levels which are prefetched.

This is to generally vet the changeset, and explore possible ways of mitigating the protocol fallback and backwards compatability.

cmeiklejohn committed:
Optimize AAE replication.

Implement a prototype optimization for active anti-entropy replication
which front-loads the first few levels of bucket retrieval through one
TCP request, instead of 1025 roundtrips.  Futher optimizations include
increasing the number of levels which are prefetched.

This is to generally vet the changeset, and explore possible ways of
mitigating the protocol fallback and backwards compatability.

cmeiklejohn commented:
cc: @Vagabond

Vagabond commented on src/riak_repl_aae_fullsync.hrl:
```diff
@@ -3,6 +3,7 @@
 -define(MSG_UPDATE_TREE, 3).
 -define(MSG_GET_AAE_BUCKET, 4).
 -define(MSG_GET_AAE_SEGMENT, 5).
+-define(MSG_GET_AAE_BUCKETS, 6).
```
Aren't you overloading message code 6 here?

cmeiklejohn replied:
Yeah, I totally did.

Vagabond commented on src/riak_repl_aae_source.erl:
```diff
@@ -280,7 +285,18 @@ key_exchange(start_key_exchange, State=#state{cluster=Cluster,
                      end;
                 (get_bucket, {L, B}) ->
-                     send_synchronous_msg(?MSG_GET_AAE_BUCKET, {L,B,IndexN}, State);
+                     lager:warning("Bucket request for ~p ~p",
```
This is presumably debugging?

Vagabond commented on src/riak_repl_aae_source.erl:
```diff
@@ -280,7 +285,18 @@ key_exchange(start_key_exchange, State=#state{cluster=Cluster,
-                     send_synchronous_msg(?MSG_GET_AAE_BUCKET, {L,B,IndexN}, State);
+                     lager:warning("Bucket request for ~p ~p",
+                                   [L, B]),
+                     case orddict:find({L, B}, Buckets) of
```
We fetch all the buckets, don't we? There should never be a cache miss here.

cmeiklejohn committed:
Remove instrumentation and cache miss handling.

Technically, there should never be a cache miss, so throw an exception
if we can not retrieve the value from the orddict.

cmeiklejohn committed:
Fix incorrect message identifier.

Vagabond commented:
+1, repl_aae_fullsync passes.

cmeiklejohn commented:
Closing; not merging.
